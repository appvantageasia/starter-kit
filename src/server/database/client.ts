import { MongoClient, MongoClientOptions } from 'mongodb';
import config from '../core/config';
import { getEncryptionKeyId } from './helpers';
import getKMS from './kms';
import buildSchemaMap from './schemaMap';

export const getRegularOptions = (): MongoClientOptions => ({
    maxPoolSize: config.db.pool,
});

export const getRegularClient = () => MongoClient.connect(config.db.uri, getRegularOptions());

export const getEncryptedClient = async (kms: Exclude<ReturnType<typeof getKMS>, null>) => {
    const keyId = await getEncryptionKeyId();

    if (!keyId) {
        throw new Error('No encryption key ID have been provided for the schema');
    }

    return MongoClient.connect(config.db.uri, {
        ...getRegularOptions(),

        // enable command monitoring for the client
        monitorCommands: true,

        autoEncryption: {
            keyVaultNamespace: kms.keyVaultNamespace,
            kmsProviders: kms.kmsProviders,
            schemaMap: buildSchemaMap(keyId),

            // configure cryptd client
            extraOptions: {
                mongocryptdURI: config.db.cryptd.uri,
                mongocryptdBypassSpawn: config.db.cryptd.mongocryptdBypassSpawn,
                mongocryptdSpawnArgs: config.db.cryptd.mongocryptdSpawnArgs,
            },
        },
    });
};
