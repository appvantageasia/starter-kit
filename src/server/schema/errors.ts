/* eslint-disable max-classes-per-file */
import { ForbiddenError, UserInputError } from 'apollo-server';

// eslint-disable-next-line import/prefer-default-export
export class InvalidInput extends UserInputError {
    constructor(errors?: any) {
        super('bad request', errors);
    }
}

export class InvalidPermission extends ForbiddenError {
    constructor() {
        super('forbidden');
    }
}
