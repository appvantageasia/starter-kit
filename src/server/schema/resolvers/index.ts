import * as enums from './enums';
import * as Mutation from './mutations';
import * as Query from './queries';
import * as Subscription from './subscriptions';
import * as types from './types';

const resolvers = {
    Query,
    Mutation,
    Subscription,
    ...types,
    ...enums,
};

export default resolvers;
