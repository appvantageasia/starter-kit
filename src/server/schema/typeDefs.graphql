schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

type AuthenticationRequiresNewPassword {
  """Token to use with the mutation `changePasswordFromAuthentication`"""
  token: String!
  """User authenticated"""
  user: User!
}

type AuthenticationRequiresTOTP {
  """Token to use with the mutation `authenticateWithTOTP`"""
  token: String!
}

union AuthenticationResponse = AuthenticationRequiresNewPassword | AuthenticationRequiresTOTP | AuthenticationSuccessful

type AuthenticationSuccessful {
  """Bearer token to use in the Authorization header"""
  token: String!
  """User authenticated"""
  user: User!
}

type AuthenticationWithWebPublicKeyCredential {
  """Allowed credentials"""
  allowCredentials: [JSONObject!]!
  """Challenge for the request"""
  challenge: String!
  """Timeout"""
  timeout: Int
  """
  Token to use with the mutation `authenticateWithWebPublicKeyCredential`
  """
  token: String!
}

type AuthenticatorSetup {
  """QRCode URI for the given authenticator secret"""
  qrcodeUri: String!
  """Authenticator secret"""
  secret: String!
}

scalar DateTime

union ExternalLink = ResetPasswordLink

scalar JSONObject

type MessageNotice {
  date: DateTime!
  message: String!
}

type Mutation {
  """Request to renew a password"""
  applyForPasswordChange(username: String!): Boolean!
  """Validate credentials (username/password) and return a Json Web Token"""
  authenticate(password: String!, username: String!): AuthenticationResponse!
  """Continue authentication with TOTP"""
  authenticateWithTOTP(password: String!, token: String!): AuthenticationResponse!
  """Authenticate with a web public key credential"""
  authenticateWithWebPublicKeyCredential(response: JSONObject!, token: String!): AuthenticationResponse!
  """Change password following a successful authentication"""
  changePasswordFromAuthentication(password: String!, token: String!): AuthenticationResponse!
  """Change password by using a token"""
  changePasswordFromToken(password: String, token: String!): Boolean!
  """Complete registration request for a web public key"""
  completeWebPublicKeyCredentialRegistration(response: JSONObject!, token: String!): Boolean!
  """Create a new account/user"""
  createAccount(email: String!, password: String!, username: String!): User!
  """Disable 2FA / Authenticator for the signed user"""
  disableAuthenticator: User!
  """Enable 2FA / Authenticator for the signed user"""
  enableAuthenticator(secret: String!, token: String!): User!
  """Generate a challenge to authenticate with web credentials"""
  generateWebCredentialAuthentication(username: String!): AuthenticationWithWebPublicKeyCredential
  """
  Take the Json Web Token (JWT) from headers and returns a new one with a renewed lifetime
  
  Authentication is required
  """
  refreshCredentials: String!
  """Create a web public key credential registration request"""
  requestWebPublicKeyCredentialRegistration: WebPublicKeyCredentialRegistrationRequest!
  """Revoke user session"""
  revokeUserSession(displayNotice: Boolean, id: ObjectID!): Boolean!
  """
  Revoke web public ket credential
  
  Return the key ID revoked
  """
  revokeWebPublicKeyCredential(id: ObjectID!): String
  """
  Update the display name for the logged in user
  
  Authentication is required
  """
  updateDisplayName(displayName: String!): User!
}

scalar ObjectID

type PaginatedUsers {
  """Number of user matching the original query"""
  count: Int!
  """User on the request page"""
  items: [User!]!
}

input Pagination {
  """Number of items to fetch from a list"""
  limit: Int!
  """Offset to apply when fetching a list"""
  offset: Int!
}

type Query {
  """
  Fetch user document for the logged in user, returns null otherwise for anonymous
  """
  currentUser: User
  """Generate challenge to authenticate with WebAuthn"""
  generateAuthenticatorChallenge(username: String!): AuthenticationWithWebPublicKeyCredential
  """Generate authenticator secret and qrcode"""
  generateAuthenticatorSetup: AuthenticatorSetup!
  """Fetch WebAuthn security keys for a username"""
  getWebauthnKeys(username: String!): [String!]!
  """List users"""
  listUsers(filter: UserFilteringRule, pagination: Pagination!, sort: UserSortingRule): PaginatedUsers!
  """Retrieve a link information"""
  retrieveLink(id: String!): ExternalLink
}

type ResetPasswordLink {
  """Token to use with mutation `changePasswordFromToken`"""
  token: String!
}

enum SortingOrder {
  """Ascending order"""
  Asc
  """Descending order"""
  Desc
}

type Subscription {
  listenSystemMessages: SystemMessage!
}

union SystemMessage = MessageNotice | UserSessionRevoked

scalar Upload

type User {
  """Public displayed on interfaces"""
  displayName: String!
  """Email address for the user"""
  email: String!
  """User object ID"""
  id: ObjectID!
  """Is the authenticator enabled for this account"""
  isAuthenticatorEnabled: Boolean!
  """Is the password expired for this account"""
  isPasswordExpired: Boolean!
  """Password expiration date"""
  passwordExpiresAt: DateTime!
  """Sessions"""
  sessions: [UserSession!]!
  """Username use for authentication"""
  username: String!
  """WebAuthn keys"""
  webAuthnKeys: [UserWebAuthnKey!]!
}

input UserFilteringRule {
  """Filter by email"""
  email: String
}

type UserSession {
  """Creation date"""
  createdAt: DateTime!
  """Expiration date"""
  expiresAt: DateTime!
  """Session ID"""
  id: ObjectID!
  """IP"""
  ip: String!
  """Last activity date"""
  lastActivityAt: DateTime!
  """User agent"""
  userAgent: String!
}

type UserSessionRevoked {
  date: DateTime!
  displayNotice: Boolean!
}

enum UserSortingField {
  """Sort by authenticator enabled state"""
  Authenticator
  """Sort by email"""
  Email
}

input UserSortingRule {
  """Field on which apply the sorting"""
  field: UserSortingField!
  """Sorting order"""
  order: SortingOrder!
}

type UserWebAuthnKey {
  """Expiration date"""
  expiresAt: DateTime!
  """Key object ID"""
  id: ObjectID!
  """Last usage date"""
  lastUsed: DateTime
  """Raw key ID"""
  rawKeyId: String!
  """User Agent"""
  userAgent: String!
}

type WebPublicKeyCredentialRegistrationRequest {
  """Options for WebAuthn"""
  options: JSONObject!
  """Token to proceed with registration"""
  token: String!
}