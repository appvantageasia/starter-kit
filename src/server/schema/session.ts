import { IncomingMessage } from 'http';
import { AuthenticationError } from 'apollo-server';
import cookie from 'cookie';
import jwt from 'jsonwebtoken';
import { getOr } from 'lodash/fp';
import { ObjectId } from 'mongodb';
import config from '../core/config';
import { generateToken, consumeToken, csrfGenerator } from '../core/tokens';
import { getDatabaseContext } from '../database';
import getIp from '../utils/getIp';

export type SessionData = {
    userId: ObjectId;
    sessionId: ObjectId;
};

export const getSessionToken = async (data: SessionData) => {
    const csrf = csrfGenerator();
    const token = generateToken('session', data, config.session.lifetime, csrf);

    // get iat and exp from token instead
    const { iat, exp } = jwt.decode(token, { json: true });

    return {
        token: generateToken('session', data, config.session.lifetime, csrf),
        csrf,
        exp: new Date(exp * 1000),
        iat: new Date(iat * 1000),
    };
};

export const readSessionToken = (token: string, csrf: string) => consumeToken<SessionData>('session', token, csrf);

export const getCSRFFromHeaders = (req: IncomingMessage): string => {
    const requestCookie = req.headers?.cookie;

    return requestCookie ? getOr('', 'CSRF', cookie.parse(requestCookie)) : '';
};

const getAuthorizationFromHeader = (req: IncomingMessage) => {
    const header = req.headers?.authorization as string;

    if (!header) {
        return null;
    }

    const [type, token] = header.split(' ');

    if (type !== 'Bearer') {
        throw new AuthenticationError('Unsupported authentication method');
    }

    if (!token) {
        throw new AuthenticationError('Missing token');
    }

    return token;
};

export const getSessionDataFromRequest = async (
    req: IncomingMessage,
    authorization?: string
): Promise<SessionData | null> => {
    const token = authorization || getAuthorizationFromHeader(req);

    if (!token) {
        return null;
    }

    const data = await (async () => {
        try {
            const csrf = getCSRFFromHeaders(req);

            return await readSessionToken(token, csrf);
        } catch (error) {
            throw new AuthenticationError('Invalid authentication token');
        }
    })();

    if (!data) {
        return null;
    }

    const { collections } = await getDatabaseContext();
    const result = await collections.userSessions.updateOne(
        { _id: data.sessionId },
        {
            $set: {
                // update user agent to keep the latest
                userAgent: req.headers['user-agent'],
                ip: getIp(req),
                // then update last activity date
                lastActivityAt: new Date(),
            },
        }
    );

    if (!result.matchedCount) {
        // session was invalidated
        throw new AuthenticationError('Session invalidated');
    }

    return data;
};
