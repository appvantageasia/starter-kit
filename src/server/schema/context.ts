import { IncomingMessage, OutgoingMessage } from 'http';
import cookie from 'cookie';
import { GraphQLFieldResolver } from 'graphql';
import { FileUpload } from 'graphql-upload';
import config from '../core/config';
import { getLanguage, getLazyTranslations, GetTranslations } from '../core/translations';
import createLoaders, { Loaders } from '../loaders';
import getIp from '../utils/getIp';
import { getCSRFFromHeaders, getSessionDataFromRequest, SessionData } from './session';
import { getLazyLoggedUser, GetLoggedUser } from './user';

export type Context = {
    ip?: string;
    language: string;
    getTranslations: GetTranslations;
    session: SessionData | null;
    getUser: GetLoggedUser;
    loaders: Loaders;
    setCSRF: (value: string) => void;
    csrf: string;
    userAgent?: string;
    origin: string;
};

export type FileUploadPromise = Promise<FileUpload>;

export type RootDocument = null;

export type RootResolver<TArgs = { [argName: string]: any }> = GraphQLFieldResolver<RootDocument, Context, TArgs>;

const createContext = async (req: IncomingMessage, res?: OutgoingMessage, authorization?: string): Promise<Context> => {
    const session = await getSessionDataFromRequest(req, authorization);
    const language = await getLanguage(req);

    return {
        ip: getIp(req),
        origin: req.headers.origin,
        language,
        session,
        getTranslations: getLazyTranslations(language),
        getUser: getLazyLoggedUser(session?.userId),
        loaders: createLoaders(),
        userAgent: req.headers['user-agent'],
        setCSRF: value => {
            if (!res) {
                throw new Error('Cannot set cookies in a call without HTTP response available');
            }

            res.setHeader(
                'Set-Cookie',
                cookie.serialize('CSRF', value, {
                    sameSite: config.cookiePolicy,
                    httpOnly: true,
                    secure: config.secureCookie,
                })
            );
        },
        csrf: getCSRFFromHeaders(req),
    };
};

export default createContext;
