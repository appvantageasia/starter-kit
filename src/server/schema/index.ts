import { makeExecutableSchema } from '@graphql-tools/schema';
import resolvers from './resolvers';
import typeDefs from './typeDefs.graphql';

const schema = makeExecutableSchema({ typeDefs: [typeDefs], resolvers });

export default schema;
