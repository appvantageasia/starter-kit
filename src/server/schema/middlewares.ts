import { AuthenticationError } from 'apollo-server';
import { GraphQLFieldResolver } from 'graphql';
import { Context } from './context';

export type GraphQLMiddleware = <TSource, TContext = Context, TArgs = { [argName: string]: any }>(
    resolver: GraphQLFieldResolver<TSource, TContext, TArgs>
) => GraphQLFieldResolver<TSource, TContext, TArgs>;

export const requiresLoggedUser =
    <TSource = any, TArgs = { [argName: string]: any }>(
        resolver: GraphQLFieldResolver<TSource, Context, TArgs>
    ): GraphQLFieldResolver<TSource, Context, TArgs> =>
    async (root, args, context, info) => {
        const user = await context.getUser();

        if (!user) {
            throw new AuthenticationError('Not authenticated');
        }

        return resolver(root, args, context, info);
    };
