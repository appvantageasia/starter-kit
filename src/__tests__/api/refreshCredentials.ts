import { gql, ApolloError } from '@apollo/client';
import { ObjectId } from 'mongodb';
import { readSessionToken } from '../../server/schema/session';
import {
    composeHandlers,
    loadFixtures,
    setupDatabase,
    cleanDatabase,
    setupWebService,
    getApolloClient,
    createSessionForUser,
} from '../helpers';
import fixtures from './authenticate.fixture.json';

const mutation = gql`
    mutation {
        refreshCredentials
    }
`;

const webService = setupWebService();

beforeEach(composeHandlers(setupDatabase, loadFixtures(fixtures), webService.initialize));

afterEach(composeHandlers(cleanDatabase, webService.cleanUp));

test('Refresh credentials requires authorizations', async () => {
    const { client } = getApolloClient(webService.url);
    const promise = client.mutate({ mutation });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Refresh credentials returns a new token with the same payload', async () => {
    const userId = new ObjectId('601f7ae763d6cfc2554f6b67');
    const { token: originalToken, csrf } = await createSessionForUser(userId);
    const { client, getCSRF } = getApolloClient(webService.url, { authorizationToken: originalToken, csrf });
    const response = await client.mutate({ mutation });
    const sessionData = await readSessionToken(response.data.refreshCredentials, getCSRF());
    expect(sessionData.userId.toHexString()).toStrictEqual(userId.toHexString());
});
