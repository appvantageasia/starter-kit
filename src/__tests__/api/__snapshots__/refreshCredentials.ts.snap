// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refresh credentials requires authorizations 1`] = `
[
  {
    "extensions": {
      "code": "UNAUTHENTICATED",
    },
    "locations": [
      {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Not authenticated",
    "path": [
      "refreshCredentials",
    ],
  },
]
`;
